platform/linux-generic/odp_crypto_ipsecmb.c

```diff
@@ -0,0 +1,1063 @@
+{
+	odp_spinlock_lock(&global->lock);
+	session->next = global->free;
+	global->free = session;
+	odp_spinlock_unlock(&global->lock);
+}
+
+static
+odp_crypto_packet_result_t *get_op_result_from_packet(odp_packet_t pkt)
+{
+	odp_packet_hdr_t *hdr = packet_hdr(pkt);
+
+	return &hdr->crypto_op_result;
+}
+
+static inline void set_crypto_op_result(odp_packet_t pkt,
+					odp_crypto_alg_err_t cipher_err,
+					odp_crypto_alg_err_t auth_err)
+{
+	odp_crypto_packet_result_t *op_result;
+
+	op_result = get_op_result_from_packet(pkt);
+	op_result->cipher_status.alg_err = cipher_err;
+	op_result->cipher_status.hw_err = ODP_CRYPTO_HW_ERR_NONE;
+	op_result->auth_status.alg_err = auth_err;
+	op_result->auth_status.hw_err = ODP_CRYPTO_HW_ERR_NONE;
+	op_result->ok = (cipher_err == ODP_CRYPTO_ALG_ERR_NONE &&
+			 auth_err   == ODP_CRYPTO_ALG_ERR_NONE);
+}
+
+static inline void set_crypto_op_result_ok(odp_packet_t pkt)
+{
+	set_crypto_op_result(pkt,
+			     ODP_CRYPTO_ALG_ERR_NONE,
+			     ODP_CRYPTO_ALG_ERR_NONE);
+}
+
+static void
+null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
+		    const odp_crypto_packet_op_param_t *param ODP_UNUSED,
+		    odp_crypto_generic_session_t *session ODP_UNUSED)
+{
+	set_crypto_op_result_ok(pkt);
+}
+
+static
+void zuc_eea3_encrypt(odp_packet_t pkt,
+		      const odp_crypto_packet_op_param_t *param,
+		      odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+
+#if ODP_DEPRECATED_API
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else {
+		set_crypto_op_result(pkt,
+				     ODP_CRYPTO_ALG_ERR_DATA_SIZE,
+				     ODP_CRYPTO_ALG_ERR_NONE);
+		return;
+	}
+#else
+	iv_ptr = param->cipher_iv_ptr;
+	_ODP_ASSERT(session->p.cipher_iv_len == 0 || iv_ptr != NULL);
+#endif
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (odp_unlikely(odp_packet_is_segmented(pkt))) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	for (uint32_t i = 0; i < MAX_BUFFERS; i++) {
+		memcpy(local.pKeys[i], session->cipher.key_data,
+		session->p.cipher_key.length);
+		memcpy(local.pSrcData[i], data, in_len);
+		memcpy(local.pIV[i], iv_ptr, session->p.cipher_iv_len);
+		local.packetLen[i] = param->cipher_range.length;
+	}
+
+	IMB_ZUC_EEA3_N_BUFFER(mb_mgr, (const void * const *)local.pKeys,
+			      (const void * const *)local.pIV,
+			      (const void * const *)local.pSrcData,
+			      (void **)local.pDstData, local.packetLen,
+			      MAX_BUFFERS);
+
+	odp_packet_copy_from_mem(pkt, in_pos, in_len, local.pDstData[0]);
+
+	set_crypto_op_result_ok(pkt);
+}
+
+static
+void zuc_eea3_decrypt(odp_packet_t pkt,
```

Comment:
This function seems to be almost identical to the encrypt function. Please try to avoid code duplication.

Reply:
Yes, there are identical. They are replaced by a common function named `zuc_eea3_cipher_op` in the latest version.